In [0]:
import requests
import re
import bs4
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np
from geopy.geocoders import Nominatim 
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

In [0]:
# Get the data from the wikibedia page
postal_codes_of_toranto_page = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(postal_codes_of_toranto_page, 'html.parser')
soup_table = soup.find_all(id="content")[0].find_all(id='bodyContent')[0].find_all('table')[0].find_all('tbody')[0]

In [64]:
# Get the data from the table into lists
postal_code = []
borough = []
neighborhood = []
for row in soup_table.children:
  if isinstance(row, bs4.element.Tag):
    postal_code.append(row.contents[1].string.strip())
    borough.append(row.contents[3].string.strip())
    neighborhood.append(row.contents[5].string.strip())
print(neighborhood)

['Neighborhood', 'Not assigned', 'Not assigned', 'Parkwoods', 'Victoria Village', 'Regent Park, Harbourfront', 'Lawrence Manor, Lawrence Heights', "Queen's Park, Ontario Provincial Government", 'Not assigned', 'Islington Avenue, Humber Valley Village', 'Malvern, Rouge', 'Not assigned', 'Don Mills', 'Parkview Hill, Woodbine Gardens', 'Garden District, Ryerson', 'Glencairn', 'Not assigned', 'Not assigned', 'West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale', 'Rouge Hill, Port Union, Highland Creek', 'Not assigned', 'Don Mills', 'Woodbine Heights', 'St. James Town', 'Humewood-Cedarvale', 'Not assigned', 'Not assigned', 'Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood', 'Guildwood, Morningside, West Hill', 'Not assigned', 'Not assigned', 'The Beaches', 'Berczy Park', 'Caledonia-Fairbanks', 'Not assigned', 'Not assigned', 'Not assigned', 'Woburn', 'Not assigned', 'Not assigned', 'Leaside', 'Central Bay Street', 'Christie', 'Not assigned', 'Not assigned', '

In [52]:
# Get the data into a pandas data frame
data = {postal_code[0]:postal_code[1:],
        borough[0]: borough[1:],
        neighborhood[0]: neighborhood[1:]}
toronto_pc_df = pd.DataFrame(data)
toronto_pc_df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [86]:
# Drop any row with no borough
rows_with_no_borough = toronto_pc_df[toronto_pc_df['Borough'] == 'Not assigned'].index
toronto_pc_df.drop(rows_with_no_borough, inplace=True)
toronto_pc_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 2 to 178
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Postal Code   103 non-null    object
 1   Borough       103 non-null    object
 2   Neighborhood  103 non-null    object
dtypes: object(3)
memory usage: 3.2+ KB


In [70]:
# Check if there are any neighborhood that is not assigned
toronto_pc_df[toronto_pc_df['Neighborhood'] == 'Not assigned']
# There are no neigborhoods that are not assigned

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

In [71]:
# The final shape of the data frame
toronto_pc_df.shape

(103, 3)

In [73]:
!wget -q -O 'toronto-geospatial-data.csv' https://cocl.us/Geospatial_data
print('Data downloaded!')

Data downloaded!


In [85]:
toronto_geospatial_data = pd.read_csv('toronto-geospatial-data.csv')
toronto_geospatial_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Postal Code  103 non-null    object 
 1   Latitude     103 non-null    float64
 2   Longitude    103 non-null    float64
dtypes: float64(2), object(1)
memory usage: 2.5+ KB


In [0]:
# Join the two data frames
toronto_df = toronto_pc_df.set_index('Postal Code').join(toronto_geospatial_data.set_index("Postal Code"))

In [109]:
toronto_df.reset_index().head()

Postal Code           Borough  ...   Latitude  Longitude
0         M3A        North York  ...  43.753259 -79.329656
1         M4A        North York  ...  43.725882 -79.315572
2         M5A  Downtown Toronto  ...  43.654260 -79.360636
3         M6A        North York  ...  43.718518 -79.464763
4         M7A  Downtown Toronto  ...  43.662301 -79.389494

[5 rows x 5 columns]

In [0]:
# This for the foursquare credentials
CLIENT_ID = "put-client-id-here"
CLIENT_SECRET = "put-client-secret"
VERSION = "20180605"
LIMIT = 100

## Exploring Neighborhoods in Toronto

This is a function that is used to get venues in a certain location, with a certain name

In [0]:
# A function that is used to get venues in a certain location, with a certain name
def get_nearby_venues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Get the data using the gathered information about the neighborhoods

In [115]:
toronto_venues = get_nearby_venues(names=toronto_df['Neighborhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [116]:
print(toronto_venues.shape)
toronto_venues.head()

(2133, 7)


Neighborhood  Neighborhood Latitude  ...  Venue Longitude     Venue Category
0         Parkwoods              43.753259  ...       -79.332140               Park
1         Parkwoods              43.753259  ...       -79.331942  Convenience Store
2         Parkwoods              43.753259  ...       -79.326351           Bus Stop
3         Parkwoods              43.753259  ...       -79.333114  Food & Drink Shop
4  Victoria Village              43.725882  ...       -79.315635       Hockey Arena

[5 rows x 7 columns]

In [118]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  ...  Venue Category
Neighborhood                                                            ...                
Agincourt                                                            4  ...               4
Alderwood, Long Branch                                              10  ...              10
Bathurst Manor, Wilson Heights, Downsview North                     23  ...              23
Bayview Village                                                      4  ...               4
Bedford Park, Lawrence Manor East                                   23  ...              23
...                                                                ...  ...             ...
Willowdale, Willowdale East                                         34  ...              34
Willowdale, Willowdale West                                          6  ...               6
Woburn                                                               5  ...               5
Woodbine Heights                                                    10  ...              10
York Mills West                                                      4  ...               4

[94 rows x 6 columns]

In [119]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 271 uniques categories.


## Analyze Each Neighborhood

Let's apply one hot encoding for the categories

In [134]:
# Perform one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# Add a new "Neighborhood" column to the one-hot encoded data
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

# Move the "Neighborhood" column to the first column
fixed_columns = ['Neighborhood'] + toronto_onehot.columns.drop(['Neighborhood']).to_list()
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Accessories Store  ...  Women's Store  Yoga Studio
0         Parkwoods                  0  ...              0            0
1         Parkwoods                  0  ...              0            0
2         Parkwoods                  0  ...              0            0
3         Parkwoods                  0  ...              0            0
4  Victoria Village                  0  ...              0            0

[5 rows x 271 columns]

In [135]:
toronto_onehot.shape

(2133, 271)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [136]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  ...  Yoga Studio
0                                         Agincourt  ...          0.0
1                            Alderwood, Long Branch  ...          0.0
2   Bathurst Manor, Wilson Heights, Downsview North  ...          0.0
3                                   Bayview Village  ...          0.0
4                 Bedford Park, Lawrence Manor East  ...          0.0
..                                              ...  ...          ...
89                      Willowdale, Willowdale East  ...          0.0
90                      Willowdale, Willowdale West  ...          0.0
91                                           Woburn  ...          0.0
92                                 Woodbine Heights  ...          0.0
93                                  York Mills West  ...          0.0

[94 rows x 271 columns]

In [191]:
toronto_grouped.shape

(94, 271)

We have 94 neighborhood, and 271 category.

The following is a function to sort the venues in a descending order and returning the most important ones.

In [0]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [193]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  ... 10th Most Common Venue
0                                        Agincourt  ...             Donut Shop
1                           Alderwood, Long Branch  ...            Yoga Studio
2  Bathurst Manor, Wilson Heights, Downsview North  ...            Supermarket
3                                  Bayview Village  ...            Yoga Studio
4                Bedford Park, Lawrence Manor East  ...           Liquor Store

[5 rows x 11 columns]

## Cluster Neighborhoods

In [194]:
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

In [185]:
set(kmeans.labels_)

{0, 1, 2, 3, 4}

In [197]:
toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

ValueError: ignored

In [199]:
# add clustering labels
# neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_.astype(int))

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Borough  ... 10th Most Common Venue
Postal Code                    ...                       
M3A                North York  ...       Doner Restaurant
M4A                North York  ...            Event Space
M5A          Downtown Toronto  ...             Restaurant
M6A                North York  ...   Ethiopian Restaurant
M7A          Downtown Toronto  ...                   Café

[5 rows x 15 columns]

Finally, let's visualize the resulting clusters

In [0]:
toronto_merged = toronto_merged[toronto_merged['Cluster Labels'] >= 1.0]

In [187]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 43.6534817, -79.3839347.


In [208]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

print("colors_array: " + str(colors_array))
print("rainbow: " + str(rainbow))

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

colors_array: [[5.00000000e-01 0.00000000e+00 1.00000000e+00 1.00000000e+00]
 [1.96078431e-03 7.09281308e-01 9.23289106e-01 1.00000000e+00]
 [5.03921569e-01 9.99981027e-01 7.04925547e-01 1.00000000e+00]
 [1.00000000e+00 7.00543038e-01 3.78411050e-01 1.00000000e+00]
 [1.00000000e+00 1.22464680e-16 6.12323400e-17 1.00000000e+00]]
rainbow: ['#8000ff', '#00b5eb', '#80ffb4', '#ffb360', '#ff0000']
